File: C++/flatten-nested-list-iterator.cpp
#include <algorithm>
#include <vector>
#include <queue>


class NestedInteger {
  public:
    // Return true if this NestedInteger holds a single integer, rather than a nested list.
    bool isInteger() const;

    // Return the single integer that this NestedInteger holds, if it holds a single integer
    // The result is undefined if this NestedInteger holds a nested list
    int getInteger() const;

    // Return the nested list that this NestedInteger holds, if it holds a nested list
    // The result is undefined if this NestedInteger holds a single integer
    const std::vector<NestedInteger> &getList() const;
};


class NestedIterator {
// Time:  O(n)
// Space: O(n)
public:
    NestedIterator(const std::vector<NestedInteger> &nestedList) {
        flatten(nestedList);
    }
    
    int next() {
        int value = data.front();
        data.pop();
        return value;
    }
    
    bool hasNext() {
        return !data.empty();
    }
private:
    std::queue<int> data;
    void flatten(const std::vector<NestedInteger> &nestedList) {
        for (const auto& element: nestedList) {
            if (element.isInteger()) {
                data.push(element.getInteger());
            } else {
                flatten(element.getList());
            }
        }
    }
};

File: Python/flatten-nested-list-iterator.py
from typing import List


class NestedInteger:
    def isInteger(self) -> bool:
        """
        @return True if this NestedInteger holds a single integer, rather than a nested list.
        """
        pass

    def getInteger(self) -> int:
        """
        @return the single integer that this NestedInteger holds, if it holds a single integer
        Return None if this NestedInteger holds a nested list
        """
        pass

    def getList(self) -> 'List[NestedInteger]':
        """
        @return the nested list that this NestedInteger holds, if it holds a nested list
        Return None if this NestedInteger holds a single integer
        """
        pass


class NestedIterator:
    # Time:  O(n)
    # Space: O(n)
    def __init__(self, nestedList: List[NestedInteger]):
        self.data = []
        self._flatten(nestedList)

    def _flatten(self, nestedList: List[NestedInteger]) -> None:
        for element in nestedList:
            if element.isInteger():
                self.data.append(element.getInteger())
            else:
                self._flatten(element.getList())

    def next(self) -> int:
        return self.data.pop(0)

    def hasNext(self) -> bool:
        return bool(self.data)
